nibabel/loadsave.py

```diff
@@ -44,6 +85,9 @@ def load(filename, **kwargs):
         raise FileNotFoundError(f"No such file or no access: '{filename}'")
     if stat_result.st_size <= 0:
         raise ImageFileError(f"Empty file: '{filename}'")
+    matches, msg = _signature_matches_extension(filename)
+    if not matches:
+        raise ImageFileError(msg)
```

Comment:
My concern here is that this adds a penalty of a double read on *good* files, which can be add up when reading many files over a network.

But I think we can use the existing machinery to avoid extra reads. Note that L94 (was L50) has a `sniff` block, which is the first N bytes of an opened file.

https://github.com/nipy/nibabel/blob/5c4f39cef1f04a9afd241242c7f0148d7e5935c9/nibabel/loadsave.py#L50

Because we don't really care about the I/O penalty of a failure, I'm okay letting the loop fail a few times before we check if it's the obvious problem.

```Diff
     sniff = None
     for image_klass in all_image_classes:
         is_valid, sniff = image_klass.path_maybe_image(filename, sniff)
         if is_valid:
             img = image_klass.from_filename(filename, **kwargs)
             return img
 
+    matches, msg = _signature_matches_extension(filename, sniff)
+    if not matches:
+        raise ImageFileError(msg)
 
     raise ImageFileError(f'Cannot work out file type of "{filename}"')
```

This would require us to add a `sniff=None` parameter to  `_signature_matches_extension` so the `open().read()` call could be skipped if `not None`.